src/ekat/kokkos/ekat_kokkos_utils.hpp

```diff
@@ -27,11 +28,10 @@ namespace ekat {
 
 namespace impl {
 
-template <typename TeamMember, typename Lambda, typename ValueType, bool Serialize>
+template <bool Serialize, typename TeamMember, typename LoopBoundaryType, typename Lambda, typename ValueType>
 static KOKKOS_INLINE_FUNCTION
 void parallel_reduce (const TeamMember& team,
-                      const int& begin,
-                      const int& end,
+                      const LoopBoundaryType& loop_boundaries,
```

Comment:
I'm not sure where this function will be used in the future, so it seemed restrictive to require `TeamThreadRange` in the non-serial loop since the lambda is user defined and I guess could be anything. Also, in my implementation (since it was my final `parallel_` call), I decided to use `ThreadVectorRange` instead of `TeamThreadRange`, so I would have needed to change the `parallel_reduce()` method anyways.

I could be misunderstanding best Kokkos practices though so I am completely ok deferring.

Reply:
So you want to support both a team and thread level || reduce? That would be nice. Have you tested both?

I'm ok with supporting both, it seems nice. What do you think about adding two "wrapper functions", called `thread_reduce` and `team_reduce`, which take only start/end, and call this function? Worth it? Not worth it?

Reply:
I like the flexibility of the mechanism and the wrapper functions idea. The wrappers would reduce the level of abstraction without reducing the ability to control things.

Reply:
Flexibility is good, but just to be clear, all TVRs have range size 1 in SCREAM params so far. The immediate need is for TTRs.

Reply:
@ambrad are you saying params are sticking with policies with vector length 1 right now?

Reply:
SCREAM and SHOC use nlev Kokkos threads on GPU and Kokkos vector size 1. They use up to (nlev + Pack::n - 1)/Pack::n Kokkos threads on CPU (but in practice just 1) and again Kokkos vector size 1. There's no advantage to Kokkos vector size > 1, and >1 would mean we'd have TVR loops inside TTR loops everywhere, which would add a lot of syntax noise.

For more, see the discussion starting "Re: Kokkos. We need to map out the || loops." on [this page](https://acme-climate.atlassian.net/wiki/spaces/NGDNA/pages/762282136/Micro-app+focused+on+P3+sedimentation).

Other parameterizations are of course free to use non-trivial TVRs. The restriction is only on routines called within the top-level SHOC and P3 kernel dispatches.

Reply:
For now, I changed back to all TTR loops. I agree that the generalization is "better" and I can implement this (tomorrow as part of this PR, or on my to-do-list for future, I'm fine with either).

Reply:
If you hard code TTR, I wonder if we could remove LoopBoundaries from the template args, and make the type of loop_boundaries to be simply `Kokkos::Impl::TeamThreadRangeBoundariesStruct<int,TeamMember>`. That would catch an error if someone tries to use the routine with a TVR, as well as it will remove a non-necessary template argument.

Reply:
Note: if you don't want to use the `Impl` namespace from Kokkos (it's always best to avoid using impl/details namespaces), we could stuff this into `ekat_kokkos_types.hpp`:

```
template<typename TeamMember>
using TeamLoopBoundaries = decltype(Kokkos::TeamThreadRange(std::declval<TeamMember>(),0,1);
```
so that you can then do
```
template<template TeamMember,...>
foo(const TeamMember& team, const TeamLoopBoundaries<TeamMember>& loop_boundaries,...);
```